tests/unit/amo/components/TestAddonReviewList.js

```diff
@@ -583,7 +583,7 @@ describe(__filename, () => {
       );
     });
 
-    it('resets the page after submitting a review', () => {
+    it.skip('resets the page after submitting a review', () => {
```

Comment:
This is the skipped test I have mentioned in the PR desc and a comment. I am not sure to understand why it is failing, but passing the `router` parameter makes it fail with a very weird message. I'll get back to that with a fresh mind, but a second brain could be useful too :)

Reply:
From IRC:

```
[20:07]  <tofumatt>	willdurand: hmmm, this failing react 16 test is weird indeed
[20:07]  <tofumatt>	I'm left scratching my head on it itoo
[20:17]  <tofumatt>	willdurand: weirdly, if I don't use shallowUntilTarget but just shallow
[20:17]  <tofumatt>	and then add the extra missing props
[20:17]  <tofumatt>	I can get it to work
```

Reply:
Changing that test to:

```jsx
it('resets the page after submitting a review', () => {
      const router = { push: sinon.stub() };
      dispatchAddon({ ...fakeAddon });
      const location = { query: { page: 2 } };
      const props = getProps({
        addon: { ...fakeAddon },
        clientApp: CLIENT_APP_FIREFOX,
        dispatch: sinon.stub(),
        errorHandler: createStubErrorHandler(),
        lang: 'en-US',
        reviews: null,
        router,
        location,
      });

      const root = shallow(<AddonReviewListBase {...props} />);

      const manager = root.find(RatingManager);
      expect(manager).toHaveProp('onReviewSubmitted');
      const onReviewSubmitted = manager.prop('onReviewSubmitted');

      // Simulate a review submission.
      onReviewSubmitted();

      sinon.assert.calledWith(router.push, {
        pathname: `/${lang}/${clientApp}${root.instance().url()}`,
        query: { page: 1 },
      });
    });
```

And adding an `import { shallow } from 'enzyme';` still runs the tests and fixes the issue. Not ideal, but it does get us to React 16... 😄

Reply:
This test case + a comment would indeed be a good thing.

I have the feeling that Enzyme 3 is not ultra stable but react 16 would be very nice.

Reply:
It would be nice to keep using `shallowUntilTarget` because I'm pretty sure this is catching a real problem with our use of `withRouter` (my comment below has details).

Reply:
Indeed, I would prefer to use `shallowUntilTarget` for the same reasons. I was hoping my patch might help expose the underlying issue, not be what we moved to ulessabdolutely necessary 😊. Your explanation below makes a lot of sense, thanks. Hopefully we can fix this and still use `shallowUntilTarget`.

Reply:
So, I ended up using `setProps()` to pass the `router` stub to the base component. It works with `shallowUntilTarget()` even though it is not ideal.

I could not use an action to change the current route because we do not store anything in the state (nothing useful I mean). It is recommended to use the `params` provided by the HOC in the doc: https://github.com/reactjs/react-router-redux#how-do-i-access-router-state-in-a-container-component, and it is also mentioned:

> You should not read the location state directly from the Redux store.

So I believe the `params` approach is fine...

`shallowUntilTarget()` (or Enzyme 3) prevents the `withRouter` HOC to pass the props correctly to the `mapStateToProps()`. Fixing the loading order of the HOC did not help 😞 

Dispatching an action to "push" a route would work only if we use the `Router` component because I think it does more than just passing the `router` to its children (which is one of the reasons we can omit Redux `Provider` component for instance).

Reply:
More info below 👇